### PR TITLE
[New Operator] - Yearn Curve vaults

### PR DESCRIPTION
**Pull request by Yashiru**
## Context 

In the [Yearn app](https://yearn.finance/#/vaults), we can find multiple types of vaults. 
Some of them are Curve vaults where you can stake Curve LP token.

This operator is focused on Yearn Curve LP token vaults.

## How the operator works
There are two steps :
- Add liquidity in the Curve pool
- Interacting with Yearn vaults

The protocols takes one of the curve pool tokens.
alEth
In this Yearn vault, it is the [alETH Curve LP token](https://etherscan.io/address/0xc4c319e2d4d66cca4464c0c2b32c9bd23ebe784e#code) that is used to be staked.

> See also [alETH Curve pool](https://etherscan.io/address/0xc4c319e2d4d66cca4464c0c2b32c9bd23ebe784e#code) smart contract

### Deposit (steps)
_Input => Token from the Curve pool_
Add liquidity in the Curve pool.
Deposit the token in the Yearn Vault.
_Output => Yearn vault token (shares)_
### Withdraw (steps)
_Input => Yearn vault token (shares)_
Withdraw from Yearn and receive the Curve LP token.
Remove liquidity (will burn the LP) and receive one of the pool token.
_Output => Token from the pool_

## Differences betweens Curve pools
Some Curve pools have different interfaces especially on the `remove_liquidity_one_coin` function.
You can find those two different function selectors:
- `remove_liquidity_one_coin(uint256,int128,uint256)`
- `remove_liquidity_one_coin(uint256,uint256,uint256)`

These differences are the consequence of the presence of functions `withdraw128()` and `withdraw256()` in the operator.

## Supported Curve pool types
- [Curve plain pools](https://curve.readthedocs.io/exchange-pools.html#plain-pools) => **100% supported**
    - Curve plain pools without ETH ✔
    - Curve plain pools with ETH ✔
- [Curve lending pools](https://curve.readthedocs.io/exchange-pools.html#plain-pools) => **partialy supported**
    - Curve plain pools without ETH ✔
    - Curve plain pools with ETH 
        - Deposit ✔
        - withdraw non ETH ✔
        - Withdraw ETH ❌
- [Curve factory - pools](https://curve.readthedocs.io/factory-pools.html) => **partialy supported**
    - Curve pool without LP token ✔
    - Metapool with LP token ❌
- [Curve metapool](https://curve.readthedocs.io/exchange-pools.html#metapools) => **not supported** ❌

### CODE REVIEW

File: contracts/libraries/YearnCurveHelpers.sol [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+import "./../interfaces/external/ICurvePool/ICurvePool.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+/// @notice Library for Yearn Curve deposit/withdraw
+library YearnCurveHelpers {
+    using SafeERC20 for IERC20;
+
+    /// @dev Get the array of token amount to send to a
+    ///      Curve 2pool to add liquidity
+    /// @param pool The curve 2pool
```
Comment (obatirou) [inline contracts/libraries/YearnCurveHelpers.sol:13]: Are pool with 2 coins named `2pool` in documentation  or is it an internal naming ?
(same question for other number of coins)

It seems we should review the naming

File: contracts/libraries/YearnCurveHelpers.sol [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    /// @param token The token to remove from the pool
+    /// @param amount The amount of token to remove from the pool
+    /// @return amounts Array of 2 token amounts sorted by Curve pool token indexes
```
Comment (obatirou) [inline contracts/libraries/YearnCurveHelpers.sol:16]: We named the return variables `amounts` but there is no mention in the code.
Should we remove the name in the natspect ?
What do you think @Yashiru ?

File: contracts/operators/Yearn/YearnCurveVaultOperator.sol [hunk 3]
```diff
@@ -0,0 +1,371 @@
...
+        }
+
+        operatorStorage.transferOwnership(msg.sender);
+
+        eth = _eth;
+        weth = IWETH(_weth);
+        withdrawer = _withdrawer;
+    }
+
+    /// @notice Add liquidity in a Curve pool that inculde ETH,
```
Comment (obatirou) [inline contracts/operators/Yearn/YearnCurveVaultOperator.sol:50]:
```suggestion
    /// @notice Add liquidity in a Curve pool that includes ETH,
```

File: contracts/operators/Yearn/YearnVaultStorage.sol [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    event VaultAdded(address vault, CurvePool pool);
+
+    /// @dev Emitted when a vault is removed
+    /// @param vault The removed vault address
+    event VaultRemoved(address vault);
+
+    /// @dev Map of vault address with underlying CurvePool
+    mapping(address => CurvePool) public vaults;
+
+    /// @notice Add a Yearn single asset vault
```
Comment (obatirou) [inline contracts/operators/Yearn/YearnVaultStorage.sol:26]: We handle only LP vaults for now

File: scripts/utils.ts [hunk 5]
```diff
@@ -289,6 +330,64 @@ export function getUniAndKncWithETHOrders(
...
+        buildOrderStruct(context.yearnVaultDepositOperatorNameBytes32, yearnVaultAddress, [
+            ["address", yearnVaultAddress],
+            ["address", tokenToDeposit],
+            ["uint256", amountToDeposit],
+            ["uint256", minVaultAmount ? minVaultAmount : 0], // 100% slippage
+        ]),
+    ];
+}
+
+// Create a Deposit order in yearn
```
Comment (obatirou) [inline scripts/utils.ts:345]: Fix those comments

File: test/unit/YearnCurveVaultOperator.unit.ts [hunk 6]
```diff
@@ -0,0 +1,770 @@
...
+            });
+
+            await context.yearnVaultStorage.connect(context.masterDeployer).removeVault(vaultToAdd);
+
+            expect(await (await context.yearnVaultStorage.vaults(vaultToAdd)).poolAddress).to.equal(ethers.constants.AddressZero);
+        });
+    });
+
+    describe("deposit()", () => {
+        it("Should revert if amount to deposit is zero", async () => { // Done
```
Comment (obatirou) [inline test/unit/YearnCurveVaultOperator.unit.ts:99]: Remove done from everywhere

File: test/unit/YearnCurveVaultOperator.unit.ts [hunk 7]
```diff
@@ -0,0 +1,770 @@
...
+                    assets: [{ token: context.yearnVaultAddresses.triCryptoVault, qty: vaultBalanceReserve }],
+                },
+            ];
+
+            const nfts = await context.nestedAssetBatcher.getNfts(context.user1.address);
+
+            expect(JSON.stringify(utils.cleanResult(nfts))).to.equal(JSON.stringify(utils.cleanResult(expectedNfts)));
+        });
+
+        it("Create with ETH and Deposit (int128) in yearn 3crypto with WETH", async () => {
```
Comment (obatirou) [inline test/unit/YearnCurveVaultOperator.unit.ts:240]: It seems to be completely identical to `Create with ETH and Deposit (uint256) in yearn 3crypto with WETH`. 
We need to fix this

File: contracts/libraries/YearnCurveHelpers.sol (hunk 1 above)
Comment (Yashiru) [reply contracts/libraries/YearnCurveHelpers.sol:13]: This naming model is not internal, it can be found in other dapps using the Curve/Ellipsis protocol.

Comment (obatirou) [reply contracts/libraries/YearnCurveHelpers.sol:13]: Alright thank you !

**Pull request closed without merging**